### ISSUE TITLE
Sortino performing worse than HODL on test data 

### ISSUE THREAD

**Issue by ktattan** [ktattan/bitcoin-trader#5]
BTC increased in value by 52% test data period (from Jan 2019 to June 2019). If our agent does not have the same or better performance as buy and hold, its useless.

Below is a plot of Close price for this period.

![Screenshot 2019-06-23 at 11 11 31](https://user-images.githubusercontent.com/17826912/59974771-d8bf3f00-95a7-11e9-8987-d3f8a943fa44.png)

**Comment by kadereub** [comment]
Interesting! 

1. Why is our Sortino ratio so much higher vs. the one  I ran (900 vs 110 thousand) but it only returns like 20% vs. 60%?

2. Is it worth testing the agents performance across different regimes, e.g. Bear, Bull markets ?

    2.1 For example bot could not return 52% in a bull market but in a bear market say where it's down 30% it is only down 10%... Then maybe willing to sacrifice that performance. 

    2.2 This probably comes down to setting a suitable benchmark, and if btc price is it then should outperform...

**Comment by ktattan** [comment]
Not sure as to 1. maybe it kept the denominator very small but simultaneously kept the numerator small (so actual returns were small but risk was tiny).

Agree with 2. We need to get some data that is a bear market that the agent hasn't been trained on (we could just reverse the current test data to make it bearish).

I got some very good results predicting 1 hour ahead using an LSTM trained on Volume and FFD of Close.

![Screenshot 2019-06-23 at 13 19 41](https://user-images.githubusercontent.com/17826912/59976253-aae2f600-95b9-11e9-8360-38bf3deb2086.png)
![Screenshot 2019-06-23 at 13 20 18](https://user-images.githubusercontent.com/17826912/59976254-aae2f600-95b9-11e9-8bce-4a9ff5a86bd4.png)

If I gave our agent an accurate 1 hour prediction as well as an ok 5 minute prediction (when we eventually train on minute data), it may do a way better job. So I think we should focus a bit more on some feature engineering for now rather than reward function (as reward function will be easy to tweak, and even David Silver said everything should be in our trining data right?):

1. LSTM prediction - what features should the LSTM be trained on? (so far I have ffd of close and volume). Maybe look at including some technical indicators e.g. RSI
2. Applying FFD to some of the existing current TA's to make them stationary

I can take 1. if you want to do 2. @kadereub ?

I think if you just read in the hourly data, apply add_indicators, then use a for loop to plot each feature, you'll easily be able to see which ones aren't stationary.

**Comment by ktattan** [comment]
Check out my lstm notebook on master branch, ive already plotted the TA's

**Comment by kadereub** [comment]
Nice! Yeah agreed need to give our RL agent more useful info...
I really want to see an agent that trades **a lot more** as in buys & sells realising profits. Basically an agent that makes money off of vol rather than the asset class direction.

Cool will take a look at the notebook, for **2**

**Comment by kadereub** [comment]
I think I know why the Sortino ratio didn't perform as well as expected, I can't read numbers!! 

It was 1 million when I ran with his Params:

> [ 2 ] Training for:  13144  time steps
> [ 2 ] Total reward:  [**1074993.5**]  (sortino)

**Comment by ktattan** [comment]
Yeah I think we need to forget about the values of the ratios and really just find out the total net worth at the very end of testing. This is a dollar amount and we can easily compare this to other benchmarks.

**Comment by kadereub** [comment]
Only the below are stationary @95% CI.
Will look at frac_diff for these and the appropriate `d` value for each.

![image](https://user-images.githubusercontent.com/31546506/59977066-99ebb200-95c4-11e9-8786-f77fc2d8ef84.png)
